File: src/model/swiglu.rs
use tch::{
    Tensor,
    nn::{self, Module, ModuleT},
};

#[derive(Debug)]
pub struct SwiGLU {
    input: nn::Linear,
    output: nn::Linear,
    dropout: f64,
}

pub fn swiglu(path: nn::Path, emb_dim: i64, hidden_dim: i64, dropout: f64) -> SwiGLU {
    SwiGLU {
        input: nn::linear(
            &path / "glu_proj",
            emb_dim,
            2 * hidden_dim,
            Default::default(),
        ),
        output: nn::linear(&path / "glu_out", hidden_dim, emb_dim, Default::default()),
        dropout,
    }
}

impl ModuleT for SwiGLU {
    fn forward_t(&self, xs: &Tensor, train: bool) -> Tensor {
        let proj = self.input.forward(xs);
        let scale = proj.narrow(-1, 0, proj.size()[proj.size().len() - 1] / 2);
        let gate = proj.narrow(
            -1,
            proj.size()[proj.size().len() - 1] / 2,
            proj.size()[proj.size().len() - 1] / 2,
        );
        let gate = &gate * gate.sigmoid();
        self.output
            .forward(&(scale * gate))
            .dropout(self.dropout, train)
    }
}

File: speckle-py/speckle/model/swiglu.py
import torch
from torch import nn
from torch.nn import functional as F


class SwiGLU(nn.Module):
    def __init__(self, emb_dim: int, hidden_dim: int, dropout: float):
        super().__init__()
        self.glu_proj = nn.Linear(emb_dim, 2 * hidden_dim)
        self.glu_out = nn.Linear(hidden_dim, emb_dim)
        self.dropout = dropout

    def forward(self, xs: torch.Tensor, train: bool) -> torch.Tensor:
        proj = self.glu_proj(xs)
        scale, gate = proj.chunk(2, dim=-1)
        gate = gate * gate.sigmoid()
        return F.dropout(self.glu_out(scale * gate), self.dropout, train)


def swiglu(emb_dim: int, hidden_dim: int, dropout: float) -> SwiGLU:
    return SwiGLU(emb_dim, hidden_dim, dropout)
